Generalised keybind code to all tiddler buttons

tadeaspaule opened the pull request:
Generalised the ```keyboard-navigation.js``` code so that all the tiddler buttons can be bound to something.

![image](https://user-images.githubusercontent.com/33228844/103480166-f8f1d000-4dd2-11eb-8eb7-d4a6389f81dc.png)

also put variables for the up/down navigation so that they can be easily rebound by people

tadeaspaule committed:
generalised keybind code to all tiddler buttons

tadeaspaule committed:
added keybinding explanation to readme

MaxGyver83 commented on README.md:
````diff
@@ -9,5 +9,13 @@ Press `Esc` to unfocus the sidebar search field.
+1. Install this plugin
+2. Refresh and find it in your Plugins tab in ControlPanel
+3. Open the keyboard-navigation.js file
+4. Near the top just edit the ```bindings``` variable for more functionality, or the ```navigate_up_key``` and ```navigate_down_key``` to change the keys used for navigation
````
I guess you meant to use single backticks:

```suggestion
4. Near the top just edit the `bindings` variable for more functionality, or the `navigate_up_key` and `navigate_down_key` to change the keys used for navigation
```

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -15,11 +15,37 @@ exports.name = "tw5-keyboard-navigation"; exports.after = ["rootwidget"];
+	clone: null,
+	'export-tiddler': null,
+	edit: 'e',
+	delete: 'd',
```
By default, this doesn't work because the delete button isn't visible. Maybe we can make it work even when the button is only visible in the dropdown menu?

And I would rather not want to delete tiddlers with a single keystroke. There is no confirmation dialog and you can't undo deleting a tiddler (afaik).

tadeaspaule replied:
OK this is a personal choice as I definitely find it useful :) but we can leave it null by default.

Regarding dialog; from my experience, if you have anything written in tiddler body, it gives you the dialog box for deleting. If it's just a name then it does it without dialog. I think this is default behavior, should be for you as well? Unless some other plugins.

Yeah I don't think there is an undo button.

Regarding visible delete button, yeah I forgot to mention this in the README (also something that gave me trouble with your original version before I figured it out), it relies on having the buttons visible, so the logic of "i have it bound so I can disable that button" actually causes you problems haha. I will put this into the README explicitly.

I took a quick look at the html, and it seems that when the buttons are hidden (in the more collapse button), they aren't present as DOM elements so we can't call ```click()``` on them

MaxGyver83 replied:
> Regarding dialog; from my experience, if you have anything written in tiddler body, it gives you the dialog box for deleting. If it's just a name then it does it without dialog. I think this is default behavior, should be for you as well? Unless some other plugins.

Oh, you are right! I just had created an empty tiddler for testing. I noticed a bug: When you press `d` for delete, then `Esc` for canceling the deletion, two tiddlers will be highlighted (in my case the one to be deleted and the next one).

> I took a quick look at the html, and it seems that when the buttons are hidden (in the more collapse button), they aren't present as DOM elements so we can't call `click()` on them

May we can simulate two clicks? One for the collapse button and another one for the delete (or whatever) button?!

tadeaspaule replied:
That's both great points, if i have more time i will mess with this, especially the more button shouldn't be difficult to do. Or go for it and do this if you have the time, would really appreciate being able to have no toolbar buttons but `more` :smile:

MaxGyver83 replied:
OK, I will take care about this.

MaxGyver83 replied:
For now, I just added a default keybinding (`m`) for the "more actions" menu. This way you can "click" every button with 1 or 2 keystrokes.

And regarding the aborted tiddler deletion: Now the next tiddler will be selected, no matter if you confirm or decline the deletion (if the last tiddler was selected, the second to last tiddler will be selected instead). That's not yet perfect, but I don't know how to detect that a deletion was canceled.

I have bumped the version to `0.6.0` and have update the [demo page](https://maximilian-schillinger.de/tw5-keyboard-navigation-plugin.html).  Thanks for your contribution!

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -15,11 +15,37 @@ exports.name = "tw5-keyboard-navigation"; exports.after = ["rootwidget"];
 // tiddler's top position should be +/- LIMIT pixels in order to be considered the topmost
 const LIMIT = 50;
 
-var ignoreKeyboardLayout = false;
```
Why do you remove this option? This is a regression.

tadeaspaule replied:
Yes sorry about this! I at first didn't see it used (as some of the e.code lines were commented out). I will bring this back, shouldn't have removed it

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -167,17 +193,12 @@ document.onkeyup = function(e) {
-	var c_released = ignoreKeyboardLayout ? (e.code == "KeyC") : (e.key == "c");
-	// e: check for ArrowUp because AltGr+E is ArrowUp in Neo2, AdNW, KOY and other keyboard layouts
-	var e_released = ignoreKeyboardLayout ? (e.code == "KeyE" && e.key != "ArrowUp") : (e.key == "e");
+	var down_key_released = e.key == navigate_up_key;
```
Here you swapped up and down. Now I go down with the "navigate_up_key" (`k`) and the other way around.

```suggestion
	var down_key_released = e.key == navigate_down_key;
```

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -167,17 +193,12 @@ document.onkeyup = function(e) {
-	// e: check for ArrowUp because AltGr+E is ArrowUp in Neo2, AdNW, KOY and other keyboard layouts
-	var e_released = ignoreKeyboardLayout ? (e.code == "KeyE" && e.key != "ArrowUp") : (e.key == "e");
+	var down_key_released = e.key == navigate_up_key;
+	var up_key_released = e.key == navigate_down_key;
```
```suggestion
	var up_key_released = e.key == navigate_up_key;
```

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -15,11 +15,37 @@ exports.name = "tw5-keyboard-navigation"; exports.after = ["rootwidget"];
+
+const BTN_CLASSNAME_PREFIX = 'tc-btn-%24%3A%2Fcore%2Fui%2FButtons%2F' // tiddler buttons have classnames in the form {BTN_CLASSNAME_PREFIX}{button name}
+const INSTANT_NAVIGATION = false; // all transitions with {navigate_up_key} and {navigate_down_key} are instant
+const INSTANT_NAVIGATION_BEFORE = 800; // if navigated before x milliseconds, transition is instant
```
Nice :+1:

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -167,17 +193,12 @@ document.onkeyup = function(e) {
-	var k_released = ignoreKeyboardLayout ? (e.code == "KeyK") : (e.key == "k");
-	var c_released = ignoreKeyboardLayout ? (e.code == "KeyC") : (e.key == "c");
-	// e: check for ArrowUp because AltGr+E is ArrowUp in Neo2, AdNW, KOY and other keyboard layouts
-	var e_released = ignoreKeyboardLayout ? (e.code == "KeyE" && e.key != "ArrowUp") : (e.key == "e");
```
Changing this is really bad for users of Neo and other modern keyboard layouts. In my layout, `d` is mapped to `e` and `AltGr` + `d` (=`e` in my layout) is mapped to the down arrow. Now, this will be interpreted as `e` (edit tiddler) although I just want to scroll down.

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -192,61 +213,56 @@ document.onkeyup = function(e) {
+				tiddler_index -= 1;
+				if (tiddler_index < 0) tiddler_index = cycle ? tiddlers.length - 1 : 0;
+			}
+			if (Date.now() - timestamp_last_navigation < INSTANT_NAVIGATION_BEFORE || INSTANT_NAVIGATION) {
```
Just for efficiency. If you want instant navigation, there is no need to check the current time on every `j`/`k` keystroke:

```suggestion
			if (INSTANT_NAVIGATION || Date.now() - timestamp_last_navigation < INSTANT_NAVIGATION_BEFORE ) {
```

MaxGyver83 reviewed (changes requested):
Thanks for your pull request! It brings good improvements. But please don't remove the features regarding alternative keyboard layouts.

tadeaspaule committed:
Update README.md

single backticks not double

Co-authored-by: Max Schillinger <maxschillinger@web.de>

tadeaspaule committed:
Update keyboard-navigation.js.tid

switch of up and down fixed

Co-authored-by: Max Schillinger <maxschillinger@web.de>

tadeaspaule committed:
addressed PR feedback
- brought back alternatekeyboardlayout support
- fixed bug with switched up and down controls
- removed default bound delete tiddler
- updated README

tadeaspaule commented:
@MaxGyver83 I think I resolved the issues you brought up, sorry for opening a PR with such botched code :disappointed: .

One question remaining though, since I don't have experience with alternate keyboard layouts; did I handle it correctly?

tadeaspaule committed:
fix used same-named var in nested find

MaxGyver83 commented on keyboard-navigation.js.tid:
```diff
@@ -15,6 +15,35 @@ exports.name = "tw5-keyboard-navigation"; exports.after = ["rootwidget"];
+	fold: null
+}
+// if you're using an alternate keyboard layout, provide a list of key codes which should be ignored
+const alternateLayoutIgnoreKeys = ["ArrowUp"]
```
That's good! :+1:

MaxGyver83 reviewed (approved):
Awesome!
